openedx/core/djangoapps/user_authn/middleware.py

```diff
@@ -0,0 +1,41 @@
+from django.conf import settings
+from django.http import QueryDict
+from django.shortcuts import redirect
+
+
+class RedirectUnauthenticatedToLoginMiddleware:
+    """
+    Middleware that redirects unathenticated users to login page.
+
+    Any GET request comming from an unathenticated user will be responded with
+    a redirect to the login url. The middleware ignores requests to login and
+    register pages.
+
+    If redirects, passes the requested url to login as 'next' query string
+    parameter.
+
+    To enable the middleware, REDIRECT_UNAUTHENTICATED_USERS_TO_LOGIN setting
+    has to be set to True.
+
+    Assumed that the requests passed to the middleware have user attribute set.
+    """
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if (
+            settings.REDIRECT_UNAUTHENTICATED_USERS_TO_LOGIN and
+            request.method == 'GET' and
+            not request.path == settings.LOGIN_URL and
+            not request.path == '/register' and
+            not request.user.is_authenticated
+        ):
```

Comment:
I'm not sure what you mean by the first point, can you explain please?
I agree with all the other points.

Reply:
@Cup0fCoffee  I am sorry if  I was not clear, what I am trying to say is unless and until  settings.REDIRECT_UNAUTHENTICATED_USERS_TO_LOGIN is evaluated to true calculating any other condition like ` request.method == 'GET'` doesn't make sense.

so the condition should be like 

```python
if settings.REDIRECT_UNAUTHENTICATED_USERS_TO_LOGIN :
# then follow all other condition check in this block 
```

This will make more sense while we read it :) hope I am making sense to you.

Reply:
@farhaanbukhsh I was just confused by "doing unnecessary calculations", since in python `and` supports short circuiting, so if the setting is `False`, the rest of the conditions wouldn't get evaluated, so it doesn't have any effect on performance.
Thank you for explaining, I agree with the argument of readability.

Reply:
I've refactored the middleware internals, let me know what you think.